### PR TITLE
[CDAP-6174] Feature/ui log viewer scrollpin

### PR DESCRIPTION
**Pull request by PGuimarais**
screen shot 2016-08-12 at 4 32 16 pm 2
- Scroll pin now tracks table position
- Adds tooltip with time when hovering over scroll pin

https://issues.cask.co/browse/CDAP-7075

Bamboo Build:
http://builds.cask.co/browse/CDAP-DRC4166

### CODE REVIEW

**Comment by ajainarayanan** [comment]
@PGuimarais missing bower.json entry. Did you do `bower install inView --save`?

**Comment by ajainarayanan** [comment]
@PGuimarais The offset at the top seems to be off..
screen shot 2016-08-03 at 6 29 30 pm

File: cdap-ui/app/directives/timeline/timeline.js [hunk 1]
```diff
@@ -229,26 +237,52 @@ function link (scope, element) {
...
       .attr('stroke', 'grey');
   }
 
-  scope.updatePinScale = function (val) {
+  scope.updatePin = function () {
+    let xPositionVal = Math.floor(xScale(scope.pinScrollingPosition));
+    if(xPositionVal < 0 || xPositionVal > maxRange){
+      return;
+    }
     if(pinHandle !== undefined){
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline.js]: This should be `typeof pinHandle !== 'undefined'`

File: cdap-ui/app/directives/timeline/timeline.js [hunk 2]
```diff
@@ -83,6 +85,7 @@ function link (scope, element) {
     scrollPinSvg = undefined;
     xAxis = undefined;
     sliderBar = undefined;
+    tooltipDiv = undefined;
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline.js]: Do we need to explicitly define the variables as `undefined`? isn't the default `undefined`? Seems to be too redundant

File: cdap-ui/app/directives/timeline/timeline.js [hunk 3]
```diff
@@ -229,26 +237,52 @@ function link (scope, element) {
...
+      .on('mouseover', function() {
+        tooltipDiv = d3.select('body').append('div')
+          .attr('class', 'tooltip')
+          .style('opacity', 0);
+
+        //Reposition tooltip if overflows on the side of the page ; tooltip width is 250
+        let overflowOffset = xScale(scope.pinScrollingPosition) + 250 > maxRange ? 250 : 0;
+
+        tooltipDiv.transition()
+                    .duration(200)
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline.js]: Indentation is missing I guess

```
tooltipDiv.transition()
  .duration(200)
  .style('opacity', 0.9)
  .attr('class', 'timeline-tooltip');
```

**Comment by ajainarayanan** [comment]
@PGuimarais For some reason we are now missing the 'Loading...' message

screen shot 2016-08-03 at 7 00 23 pm

This is happening while the logs are being fetched/rendered in UI.

**Comment by elfenheart** [comment]
bower component is not in bower.json

File: cdap-ui/bower.json [hunk 4]
```diff
@@ -44,7 +44,9 @@
     "d3-tip": "0.6.7",
     "esprima": "2.0.0",
     "ngInfiniteScroll": "1.2.1",
-    "d3-timeline": "0.0.5"
+    "d3-timeline": "0.0.5",
+    "inView": "^1.0.0",
+    "angular-inview": "^1.5.7"
```
Comment (elfenheart) [inline cdap-ui/bower.json]: use exact versions for these. We don't want variable version

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 5]
```diff
@@ -82,8 +82,35 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+    //Top of the table
+    if(index === 0 && isInview){
+      viewMap = {};
+      viewMap[index] = isInview;
+      newTime = this.displayData[index].log.timestamp;
+      this.updateScrollPositionInStore(newTime);
+    }
+    else {
+      //This is the top of the table
+      if(isInview && (viewMap[index-1] === false || viewMap[index-1] === undefined)){
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: checking undefined is wrong

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 6]
```diff
@@ -70,8 +71,9 @@ function TimelineController ($scope, LogViewerStore, LOGVIEWERSTORE_ACTIONS, myL
 
   LogViewerStore.subscribe(() => {
     this.pinScrollPosition = LogViewerStore.getState().scrollPosition;
-    if($scope.updatePinScale !== undefined){
-      $scope.updatePinScale(this.pinScrollPosition);
+    if($scope.updatePin !== undefined){
```
Comment (elfenheart) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: will throw an error if undefined

**Comment by elfenheart** [comment]
the pin does not move at all for me

**Comment by elfenheart** [comment]
timeline is incorrect

screen shot 2016-08-08 at 1 43 55 pm

**Comment by elfenheart** [comment]
screen shot 2016-08-08 at 4 46 55 pm

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 7]
```diff
@@ -54,6 +54,7 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
         break;
     }
   };
+  this.page = angular.element(window);
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: does this need to be in a public variable?

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 8]
```diff
@@ -184,8 +185,38 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
   let numEvents = 0;
   this.toggleExpandAll = false;
 
-  let unsub = LogViewerStore.subscribe(() => {
+  let proximityVal = Number.MAX_VALUE;
+  let newTime;
+
+  this.inViewScrollUpdate = (index, isInview, event) => {
+
+      if(isInview) {
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js:217]: you should add comments explaining all the calculations here

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 9]
```diff
@@ -505,11 +536,21 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
         if(res.length === 0){
           this.renderData();
           getStatus();
+          if(!this.applicationIsRunning){
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: this variable is never set or used anywhere

File: cdap-ui/bower.json [hunk 10]
```diff
@@ -44,7 +44,9 @@
     "d3-tip": "0.6.7",
     "esprima": "2.0.0",
     "ngInfiniteScroll": "1.2.1",
-    "d3-timeline": "0.0.5"
+    "d3-timeline": "0.0.5",
+    "inView": "1.0.0",
```
Comment (elfenheart) [inline cdap-ui/bower.json]: does inView comes prepackaged in angular-inview? we are not including inView in gulp file, so it doesn't seem like this dependency is needed?

**Comment by elfenheart** [comment]
I think when you shift the start time, the scroll pin and the scroll position should go to the start of the log

**Comment by elfenheart** [comment]
screen shot 2016-08-12 at 7 04 12 pm

Sometimes this would happen..

**Comment by elfenheart** [comment]
There is still the problem that clicking on the scroll-pin and dragging it will drag the blue start point as well

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 11]
```diff
@@ -520,13 +566,25 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
           return;
         }
 
         this.renderData();
+        //Update the scroll needle to be positioned at the first element in the rendered data
+        if(this.displayData.length > 0){
+          this.updateScrollPositionInStore(this.displayData[0].log.timestamp);
+        }
+
+        this.cacheSize = res.length - this.cacheDecrement;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: where is cacheDecrement coming from?

Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: actually, is this variable being used at all?

File: cdap-ui/app/directives/timeline/timeline.js [hunk 12]
```diff
@@ -455,42 +464,25 @@ function link (scope, element) {
 
         for(var verticalStackSize = 4; verticalStackSize >= 0 && (errorCount > 0 || warningCount > 0); verticalStackSize--){
           if(errorCount > 0){
-            if(eventCount >= 5){
               timescaleSvg.append('circle')
                 .attr('cx', xScale(keyThree) + handleWidth)
-                .attr('cy', (verticalStackSize+1) * 7)
+                .attr('cy', (verticalStackSize+1) * 7 - 2)
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline.js:470]: we could add a comment for `-2`

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 13]
```diff
@@ -433,20 +437,22 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
-      this.renderData();
-    }
+      if(res.length > 0){
+        this.fromOffset = res[res.length-1].offset;
+
+        angular.forEach(res, (element, index) => {
+          //Format dates properly for rendering and computing
+          let formattedDate = new Date(res[index].log.timestamp);
+          res[index].log.timestamp = formattedDate;
+          res[index].log.displayTime = ((formattedDate.getMonth() + 1) + '/' + formattedDate.getDate() + '/' + formattedDate.getFullYear() + ' ' + formattedDate.getHours() + ':' + ((formattedDate.getMinutes()<10) ? '0'+formattedDate.getMinutes() : formattedDate.getMinutes()) + ':' + formattedDate.getSeconds());
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Can we try `$moment` here?

**Comment by ajainarayanan** [comment]
👍 LGTM

**Pull request merged**